File: apps/web/src/lib/admin/operations.ts
import { adminFetch } from './api';

export type JobStatus =
  | 'pending'
  | 'running'
  | 'success'
  | 'failed'
  | 'timeout';

export interface ParamsSchema {
  title?: string;
  description?: string;
  properties?: Record<string, unknown>;
  required?: string[];
  [key: string]: unknown;
}

export interface CommandCatalogEntry {
  command: string;
  label: string;
  description: string;
  timeout: number;
  params_schema: ParamsSchema;
}

export interface AdminJob {
  id: number;
  command: string;
  params_json: Record<string, unknown> | null;
  status: JobStatus;
  result_json: Record<string, unknown> | null;
  stderr_tail: string | null;
  started_at: string | null;
  finished_at: string | null;
  admin_id: string;
  admin_email: string | null;
  created_at: string;
}

export interface JobsPage {
  items: AdminJob[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export async function getCatalog(): Promise<CommandCatalogEntry[]> {
  const response = await adminFetch('/operations/catalog');
  if (!response.ok) {
    throw new Error(`catalog_failed_${response.status}`);
  }
  return (await response.json()) as CommandCatalogEntry[];
}

export async function runCommand(
  command: string,
  params?: Record<string, unknown> | null,
): Promise<AdminJob> {
  const response = await adminFetch('/operations/run', {
    method: 'POST',
    body: JSON.stringify({ command, params: params ?? null }),
  });
  if (!response.ok) {
    if (response.status === 400) throw new Error('command_not_allowed');
    if (response.status === 422) {
      const detail = (await response.json().catch(() => ({}))) as unknown;
      throw new Error(`invalid_params:${JSON.stringify(detail)}`);
    }
    throw new Error(`run_failed_${response.status}`);
  }
  return (await response.json()) as AdminJob;
}

export interface ListJobsParams {
  page?: number;
  size?: number;
  status?: JobStatus | '';
}

export async function listJobs(
  params: ListJobsParams = {},
): Promise<JobsPage> {
  const search = new URLSearchParams();
  if (params.page !== undefined) search.set('page', String(params.page));
  if (params.size !== undefined) search.set('size', String(params.size));
  if (params.status) search.set('status', params.status);
  const qs = search.toString();
  const path = qs ? `/operations/jobs?${qs}` : '/operations/jobs';
  const response = await adminFetch(path);
  if (!response.ok) {
    throw new Error(`list_jobs_failed_${response.status}`);
  }
  return (await response.json()) as JobsPage;
}

export async function getJob(id: number): Promise<AdminJob> {
  const response = await adminFetch(`/operations/jobs/${id}`);
  if (!response.ok) {
    if (response.status === 404) throw new Error('not_found');
    throw new Error(`get_job_failed_${response.status}`);
  }
  return (await response.json()) as AdminJob;
}

File: apps/api/app/api/v1/admin/operations.py
from __future__ import annotations

from typing import Annotated, Literal

from fastapi import APIRouter, HTTPException, Query, status

from app.api.deps import AdminDep, SessionDep
from app.core.logging import get_logger
from app.repos import admin_jobs as jobs_repo
from app.schemas.admin_jobs import (
    CatalogEntry,
    JobCreateRequest,
    JobListPage,
    JobOut,
)
from app.services.admin import operations as ops_service
from app.services.admin.operations import (
    CommandNotAllowedError,
    InvalidParamsError,
)
from app.services.admin.operations_catalog import ALLOWED_COMMANDS

router = APIRouter(prefix="/operations", tags=["admin-operations"])
log = get_logger("app.admin.operations")


@router.get("/catalog", response_model=list[CatalogEntry])
async def get_catalog(
    admin: AdminDep,  # noqa: ARG001
) -> list[CatalogEntry]:
    return [
        CatalogEntry(
            command=key,
            label=spec["label"],
            description=spec["description"],
            timeout=int(spec["timeout"]),
            params_schema=spec["params_model"].model_json_schema(),
        )
        for key, spec in ALLOWED_COMMANDS.items()
    ]


@router.post(
    "/run",
    response_model=JobOut,
    status_code=status.HTTP_201_CREATED,
)
async def run_command(
    body: JobCreateRequest,
    admin: AdminDep,
    session: SessionDep,
) -> JobOut:
    try:
        job = await ops_service.enqueue_job(
            session,
            admin_id=admin.id,
            command=body.command,
            raw_params=body.params,
        )
    except CommandNotAllowedError as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"command_not_allowed: {exc}",
        ) from exc
    except InvalidParamsError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY,
            detail=f"invalid_params: {exc}",
        ) from exc

    log.info(
        "admin_job_enqueued",
        admin_id=str(admin.id),
        command=body.command,
        job_id=job["id"],
    )
    job["admin_email"] = admin.email
    return JobOut(**job)


@router.get("/jobs", response_model=JobListPage)
async def list_jobs(
    admin: AdminDep,  # noqa: ARG001
    session: SessionDep,
    status_filter: Annotated[
        Literal["pending", "running", "success", "failed", "timeout"]
        | None,
        Query(alias="status"),
    ] = None,
    page: Annotated[int, Query(ge=1)] = 1,
    size: Annotated[int, Query(ge=1, le=100)] = 20,
) -> JobListPage:
    items, total, pages = await jobs_repo.list_jobs(
        session, page=page, size=size, status=status_filter,
    )
    return JobListPage(
        items=[JobOut(**item) for item in items],
        total=total,
        page=page,
        size=size,
        pages=pages,
    )


@router.get("/jobs/{job_id}", response_model=JobOut)
async def get_job(
    job_id: int,
    admin: AdminDep,  # noqa: ARG001
    session: SessionDep,
) -> JobOut:
    job = await jobs_repo.get_job(session, job_id)
    if job is None:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="job_not_found",
        )
    return JobOut(**job)
